### DISCUSSION TITLE
BTT Octopus Pro F446 build

### DISCUSSION THREAD

**Discussion by yoft** [grblHAL/STM32F4xx#141] [Q&A]
Hi,
I've been attempting to a build for the BTT Octopus Pro F446
https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-Pro

It has a 32kb bootloader and 12MHz crystal.

I've written the board map without any issues. However, I'm having trouble with the build configuration.
I've never worked on linker scripts before so I was wondering if you had any guidance with this?

Thanks!

**Comment by terjeio** [comment]
Using Eclipse to build?
Perhaps [this lnker](https://github.com/grblHAL/STM32F4xx/blob/master/STM32F446RETX_BL64K_NONVS_FLASH.ld) script can be used, it assumes a 64K bootloader and no EEPROM emulation in flash. It has a 32Kbit EEPROM that should be used for settings storage.

**Reply by yoft** [reply]
Ah! I missed the external eeprom, that was definitely causing me grief.
As for the linker script, do I need to adjust the bootloader size to 32kb?

Thanks again!

**Reply by terjeio** [reply]
> do I need to adjust the bootloader size to 32kb?

I do not know, best to to try first with 64K if you do not know its size. I assume you are programming by using a SD card image?

**Reply by yoft** [reply]
I think it is 32kb, but I'll try both and let you know.
Yes, so far popping firmware.bin on the SD card has worked fine (gets renamed to firmware.cur) but I haven't been able to communicate with the board over usb or serial after updating.

**Reply by yoft** [reply]
Hi,
I've tried a few variations on the build, the latest is here. I'm still unable to communicate after flashing. I tried both 64 and 32k bootloader sizes and  added #define EEPROM_ENABLE 32 to the board map.

Here is the latest attempt:
https://github.com/yoft/STM32F4xx/tree/new-octo

Thank you aain for your help!

**Reply by terjeio** [reply]
Communication via USB can be tricky to get right since the clock supplied to the USB peripheral has to be exactly 48 MHz. The clock setup [differs between F446 variants](https://github.com/yoft/STM32F4xx/blob/89df5e5025eba7bd3dbb9f9ff3dfc30cd63bbd13/Src/main.c#L60C1-L132C8) making it even more tricky.

Try changing [this line](https://github.com/yoft/STM32F4xx/blob/89df5e5025eba7bd3dbb9f9ff3dfc30cd63bbd13/Src/main.c#L113) to:
`  #elif defined(BOARD_FYSETC_S6) || defined(BOARD_BTT_OCTOPUS_PRO)`

Regardless of this you can measure the voltage on a stepper enable pin: removing a stepper driver from its socket and measure in the socket is IMO the easiest/safest. It should be 3.3V (or 5V if the signal is buffered) if the code was sucessfully started by the bootloader. Do this to determine the size of the bootloader.

**Reply by yoft** [reply]
Hi @terjeio,
I've tried your suggestions with no success. Both 32k and 64k bootloaders had 5v on the enable output.

I've found some settings used for this board from another source [here](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/tree/9675aed987a4ab288e3b097e09f456dc81fdb38d/Firmware/RepRapFirmware/RRFBuild-SourceCode/CoreN2G/src/STM32F4/variants/STM32F446)

I've copied them into [my repo](https://github.com/yoft/STM32F4xx/tree/new-octo)

I'm seeing 5v on driver0 Enable output on the new linker script committed to my repo too. Is it possible this is a feature of the bootloader?

I've also tired setting
/#define USB_SERIAL_CDC 0
and using a usb/serial convertor @ 115200 baud and have had no response to that either. I tried installing the [Klipper firmware](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/tree/9675aed987a4ab288e3b097e09f456dc81fdb38d/Firmware/Klipper) however I've never used this before. When using the UART variant, it was periodically sent a weird character over serial every few seconds. When using the USB variant, it showed up in device manager as an Unknown Device, which is more than I had before.

I'm very grateful for your help in this

**Reply by terjeio** [reply]
> I've found some settings used for this board from another source [here](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/tree/9675aed987a4ab288e3b097e09f456dc81fdb38d/Firmware/RepRapFirmware/RRFBuild-SourceCode/CoreN2G/src/STM32F4/variants/STM32F446)

Ok, it seems the bootloader is 32K from that:

`FLASH (rx)      : ORIGIN = 0x8008000, LENGTH = 512K - 32K`

Have you tried with BTTs [clock setup](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/blob/9675aed987a4ab288e3b097e09f456dc81fdb38d/Firmware/RepRapFirmware/RRFBuild-SourceCode/CoreN2G/src/STM32F4/variants/STM32F446/variant.cpp#L224-L255)?

And if you have a scope you can check the UART TX line, is there traffic there on boot and is the bit rate 115200 baud (~8.7 microsecond/bit)?

**Reply by yoft** [reply]
Hi @terjeio,
I've made some changes and grabbed a linker script from the BTT repo.
[Here](https://github.com/grblHAL/STM32F4xx/compare/master...yoft:STM32F4xx:new-octo?expand=1) Are the changes.
And [here](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/blob/master/Firmware/RepRapFirmware/RRFBuild-SourceCode/CoreN2G/src/STM32F4/variants/STM32F446/variant.cpp) is where I got the settings from.

The driver enable output appears to be 5v no matter which linkerscript I use. I've tried the first two driver enable outputs.

Compiler preprocessor symbols:
USE_HAL_DRIVER
BOARD_BTT_OCTOPUS_PRO
STM32F446xx
HSE_VALUE=12000000
_VOLUMES=1
_USE_IOCTL=1
_USE_WRITE=1

If you have any further ideas, I'd love to hear them.
Thank you again for your help.

**Comment by terjeio** [comment]
Do not enable the EEPROM until you have communication working. And disable flash emulation as well by leaving only line 2663 in [this block of code](https://github.com/grblHAL/STM32F4xx/blob/d65df3051cedda74ba2140f189d334dfc3f9d88e/Src/driver.c#L2656-L2664).
If disabling USB CDC is the UART still dead? Do you have access to an oscilloscope for checking the TX line? Or are there LEDs on the USB <> UART breakout? If so is the TX LED briefly flashing on startup?

To 100% verify that the code has started you could try to continously toggle a GPIO pin early in the the startup sequence, after outputs has been initialized - before [this line](https://github.com/grblHAL/STM32F4xx/blob/d65df3051cedda74ba2140f189d334dfc3f9d88e/Src/driver.c#L2263). This to determine if the clock tree is incorrectly setup or if the booloader/loader script combo is wrong.

**Reply by yoft** [reply]
I've tried disabling EEPROM and flash, however I receive a few errors and the build fails.
`/../Src/flash.c:44: undefined reference to '_EEPROM_Emul_Start'`
I'm confused by this as FLASH_ENABLE = 0 so flash.h should not be included.

**Reply by terjeio** [reply]
Try with this: 
[flash.zip](https://github.com/grblHAL/STM32F4xx/files/12924894/flash.zip)

**Comment by yoft** [comment]
Hi @terjeio,
Good news!
I went back to basics and learnt how to write a simple application that blinks an output.
Whilst doing this, and using other outputs to indicate where errors had occurred I believe I found the issues!
The main issue is that it threw an error when attempting to configure PLL, I suspect that the bootloader configures the system clock to use PLL. So I've told it to use HSI, configure PLL , then use PLL again.
As soon as this was fixed, I was able to connect to the board through USB using IoSender! Huge relief! 

Now just to sort out the EEPROM connection issues.

**Reply by terjeio** [reply]
Good to hear!

> Now just to sort out the EEPROM connection issues.

Setting EEPROM_ENABLE to 32 and I2C_PORT to 1 isn't working?

**Reply by yoft** [reply]
Hi @terjeio ,
Sorry I didn't get back to you sooner.
The day after my previous message I tried compiling again and had no eeprom issues!
I've since been working on getting a second hand VFD connected over modbus, it's an Omron V7 which uses "memobus" (modbus with a little more data). I was able to modify the modvfd to get it going, after a day of frustration.

I'll tidy up everything I've got working and submit a pull request when possible.

Thank you for your help and continual hard work!

**Reply by terjeio** [reply]
I am not in a hurry so take your time.
Please split pull requests for the driver and spindle code, better to review them separately.

**Reply by dantetango** [reply]
Any chance of an update? I have a spare octopus pro and I'd like to replace my woodpecker board.

**Reply by yoft** [reply]
Hi @dantetango ,
Progress is [here](https://github.com/grblHAL/STM32F4xx/discussions/166#discussioncomment-8938914).
I haven't had time to put together a pull request yet, hoping someone else can!

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
@dantetango I think I can help you with this. You can have a look at [my fork](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx). There the BTT  Octopus Pro is already included. However, it is not tested, because I only own a BTT Octopus 1.1. Just let me know in case you need any help.

**Reply by dantetango** [reply]
> @dantetango I think I can help you with this. You can have a look at [my fork](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx). There the BTT  Octopus Pro is already included. However, it is not tested, because I only own a BTT Octopus 1.1. Just let me know in case you need any help.

Right now I'm away from the project for the next week, so I'm in no rush. It's going on a 3018, just replacing the cheap AliExpress board because I have an octopus laying around. No real reason, just because. Grblhal, also "just because". 

As it stands, what's the process for building off your fork? I'm only familiar with the web builder

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
There are multiple ways how you can build the firmware. However, I am using PlatformIO. There is a how-to in the README file of the repo. If you are familiar with Visual Studio Code I recommend you to use the [PlatformIO extension](https://platformio.org/install/ide?install=vscode).

Prior to building you have to customize the map file, so you can specify which pins are used for what.

> If you want to customize the build, take a look at [platformio.ini](https://github.com/grblHAL/STM32F4xx/blob/master/platformio.ini) where you can uncomment/customize additional build_flags and lib_deps in the [common] or [env:some_board_variant] sections.

> Also see [Inc/my_machine.h](https://github.com/grblHAL/STM32F4xx/blob/master/Inc/my_machine.h) but beware that settings in this file are ignored by PlatformIO (via the -D OVERRIDE_MY_MACHINE= build flag). If they're not already present in the platformio.ini file, you will need to add them to the build_flags option in the appropriate section.


However, I have one question: which processor is mounted on your BTT Octopus Pro? Is it the "STM32F446" or the "STM32F429" or the "STM32H723"? Because this code works ONLY with the "STM32F446" variant!!

**Reply by dantetango** [reply]
No experience with it yet but at a glance it looks straightforward enough.

And yes, I have the former with 2209's

**Reply by yoft** [reply]
Hey @ErdbeermarmeladebrotmitHonig ,
Did you ever end up putting in the pull request?
I've been asked in another thread about the F429 and was wondering if you could throw in the code for that? It will obviously still need a linker script, but if the clock setting are in that's a start.
[Clock settings here](https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/blob/master/Firmware%2FMarlin-bugfix-2.0.9.3.x%2Fbuildroot%2Fshare%2FPlatformIO%2Fvariants%2FMARLIN_F429ZG%2Fvariant.cpp)

Thanks!

**Reply by ashokmanker** [reply]
@ErdbeermarmeladebrotmitHonig 
I have btt_octopus_pro with 429 chip. What should I define in this board: ?

"board = genericSTM32F407VGT6
board_build.ldscript = STM32F429VGTX_BL32K_FLASH.ld"

ldscript is stm32f429 is available in repo. What should I define in board?

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Servus,

unfortunately I wasn't even able to test the code located in my repo on my board.
@ashokmanker there is a [migration guide](https://www.st.com/resource/en/application_note/an4658-migration-of-applications-from-stm32f429439-lines-to-stm32f446-line-stmicroelectronics.pdf) from STM32F429 to STM32F446. It looks like there is not that much of a difference between these two chips (in terms of pin assignment).

I don't know if you are familiar with coding. If not, you can give it a try to use the code from [my fork](https://github.com/ErdbeermarmeladebrotmitHonig/STM32F4xx) and define BOARD_BTT_OCTOPUS_1_1. However, you must change the the linker script in the platform.ini file to the STM32F429 chip. Keep in mind, that this linker script will not work, because _FLASH_VectorTable = ORIGIN(FLASH); is not defined. In addition I never brought the EEPROM Emul to work (it doesn't really matter because the Octopus board has EEPROM onboard).

**Reply by ashokmanker** [reply]
Today I wanted to try your build. But unfortunately platformio stuck to initializing, nothing happened. I tried solution on net. But didn't find. Now reinstalling whole vscode and platformio.

About flash vector table. Link given by @yoft may be useful.

Solved platformio problem.
Compiling for octopus_pro and 1.1 gives error in diskio.c

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
I just updated my repo, now it contains a build config for STM32F429 for BTT Octopus 1.1 and Pro. For me it builds, however I am not able to test this code as I do not own such a board.

**Reply by ashokmanker** [reply]
Thank Bro,
Its compiling successfully but its HSE should be 8000000, But you have defined 12000000. I changed this then I got several errors.

Compiled with HSE 8000000, with commenting out error in btt_octopus_1_1.h file.
USB port is not visible in windows. How to verify its working?

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Ok I am sorry, I messed up, the previous version was not working correctly ... :D
I took the wrong platformio board ...
I just updated my code again. In addition you have to add the right chip in the platform.h file in the grbl submodule. You can just copy the following stuff, this should work.

```
#if defined(STM32F103xB) || defined(STM32F103xE) || defined(STM32F401xC) || defined(STM32F401xE) || defined(STM32F407xx) || defined(STM32F411xE) || \
    defined(STM32F412Vx) || defined(STM32F429xx) || defined(STM32F446xx) || defined(STM32F756xx) || defined(STM32F765xx) || defined(STM32H743xx) || defined(STM32H723xx)
#define STM32_PLATFORM
#endif
```

**Reply by ashokmanker** [reply]
compiled and flashed with TMC and ganged axis disabled. I have TMC5160 and its not compiling with that.
flashed firmware.bin via sdcard. but usb port not detected. So how to verify that image is working?

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Ok I haven't changed the clock settings. Now, I added the right clock settings. However, I don't know if they will work.
I had the same problem, that the board didn't show up in my device manager, this is a problem with the clock settings.

**Reply by ashokmanker** [reply]
Thanks I will check and update

**Reply by yoft** [reply]
I did have a problem where I think the bootloader was changing the clock settings, so I had to change them twice. An easy way to check is to put an led blink script right at the start of the program after clock setup. So nothing else launches and the led blinks instead.

**Reply by ashokmanker** [reply]
Still not detecting USB port on PC.

**Reply by yoft** [reply]
Did you try a blink script? If so, that will stop the usb from working, but will confirm the linker script is correct.

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
I just updated the linker script, maybe there was one faulty region, I removed it.

**Reply by ashokmanker** [reply]
Still not detecting on PC. May be ld script is for 429VGTX, but processor in Octopus_pro is 429ZGTX. Thats also seems correct accroding to link shared by @yoft [https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/blob/master/Firmware%2FMarlin-bugfix-2.0.9.3.x%2Fbuildroot%2Fshare%2FPlatformIO%2Fvariants%2FMARLIN_F429ZG%2Fvariant.cpp](url)

**Reply by ashokmanker** [reply]
Any progress in this direction?

**Reply by ErdbeermarmeladebrotmitHonig** [reply]
Not from my side, as I am not working on it. I can't debug anything as I don't own any board with a STM32F429 chip. You mentioned that the linker script is for STM32F429VGTX but the BTT Octopus has a STM32429ZGTX, but in my opinion this should not make that much of a difference. However, you could try the linker script from Marlin, maybe it works with that one. Keep in mind, that you have to add the "_FLASH_VectorTable".
If it is not working with this linker script you could do what @yoft mentioned: try a blink script.

**Reply by ashokmanker** [reply]
Amyone else helping working 429 board
